src/tools/interop/idt/utils/shell.py

```diff
@@ -0,0 +1,122 @@
+#
+#    Copyright (c) 2023 Project CHIP Authors
+#    All rights reserved.
+#
+#    Licensed under the Apache License, Version 2.0 (the "License");
+#    you may not use this file except in compliance with the License.
+#    You may obtain a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+#
+
+import multiprocessing
+import shlex
+import subprocess
+
+import psutil
+
+from . import log
+
+logger = log.get_logger(__file__)
+
+
+class Bash:
```

Comment:
The heavy handed recursive kill on stop cleans up everything if the program stops gracefully.

For CTRL+C, the signal gets propagated to subprocs, so long as they actually exit there (tcpdump and adb do).

```
2023-11-05 14:08:09,254 WARNING {pcap} [start_pcap] Retrying pcap with sudo...
tcpdump: data link type PKTAP
tcpdump: listening on any, link-type PKTAP (Apple DLT_PKTAP), snapshot length 524288 bytes
2023-11-05 14:08:11,258 INFO {pcap} [start_pcap] Pcap output path /Users/bozowski/idt/connectedhomeip/src/tools/interop/idt/IDT_ARTIFACTS/idt_20231105_140807/pcap/idt_20231105_140807_pcap.pcap

_____________________________
Initializing platform Android
_____________________________

2023-11-05 14:08:11,476 ERROR {android} [_device_id_user_input] Connect additional android devices via USB and press enter OR
2023-11-05 14:08:11,476 ERROR {android} [_device_id_user_input] Enter (copy paste) the target device id from the list of available devices below OR
2023-11-05 14:08:11,476 ERROR {android} [_device_id_user_input] Enter $IP4:$PORT to connect wireless debugging.
^C694 packets captured
714 packets received by filter
0 packets dropped by kernel
```

Agreed though it would be ideal to have a single cleanup function in the main loop that takes care of missed calls to stop etc. and uses a more elegant abstraction like process group.

Reply:
I think we're OK on this for practical concerns right now, but obviously I'm quite bothered by the same topic (the scripts to check/kill children, the cleanup from root proc that was previously in here etc) so I'm more than willing to revisit this. Will include this in follow up top comment in a moment.